### ISSUE TITLE
FR: Support UnifiedPush for Syncing Changes to Android in a More Battery Saving Way

### ISSUE THREAD

**Issue by Zocker1999NET** [syncthing/syncthing#9454]
### Feature description

(I open this issue on the main repo rather on the Android wrapper as I acknowledge that this feature most probably requires changes to the protocol and so the main daemon itself.)

In short, this is a reopening of syncthing/syncthing-android#230 but by basing it on the [UnifiedPush specification](https://unifiedpush.org/developers/spec/) instead of Google Cloud Messaging.

### UnifiedPush summary

[UnifiedPush](https://unifiedpush.org/) (UP) is a set of specifications allowing apps to receive push notifications in a way the user chooses. Users are required to install & configure a distributor app (e.g. [ntfy](https://unifiedpush.org/users/distributors/ntfy/) or [UP-FCM Distributor](https://unifiedpush.org/users/distributors/fcm/)). [Other apps](https://unifiedpush.org/users/apps/) can use broadcast intents to then enable UP for themselves.

TL;DR: On request via broadcast intents, an app gets assigned a unique HTTP(S)-URL by the UP distributor. The app can then e.g. forward this URL to its upstream service, *similar* to a GCM token. If the upstream service then wants to push something to the app, the service POSTs on said URL. The data which gets POSTed is then forwarded 1:1 by the distributor (using its app) to the app. This happens using a broadcast intent which should wake up the app (assuming Android’s native or vendor optimizations do not hinder that, but Syncthing has the same problem now nontheless).

### Implementation Idea

This idea may be implemented as follows:

1. When pairing an Android device to another device (any OS), these devices start to connect to each other as of now.
2. Over this connection, when UP is enabled, the Android device declares its use of UP and forwards a UP-URL to the other device.
3. When the other device wants to sync something to/from the Android device, it can utilize UP to reach it.
4. The Android device wakes up and establishes a connection to other device as usual (see points below for technicalities), initiating a sync.

Points which IMO are open for discussion:

- Should Syncthing first try to find the Android device through other means (local/global discovery, direct connection, …) or should UP be the first method of discovery?
  - On a protocol level, the order of the discovery methods should be defined by the Android device as that may depend on its configuration.
  - This may be user-configurable.
  - The default behavior may be changed over time, depending on the stability of UP and continuation of discovery (point below).
- How UP-initiated discovery continues until a direct connection is established?
  - The UP message can contain information to speed up connection establishment.
  - As I understand the Syncthing protocol and assuming strict NATs & firewalls on both sides, global discovery is still required per se.
  - Can a unidirectional message be enough to decide if both devices may be able to establish a direct connection using local discovery?

### Problem or use case

Implementing UnifiedPush support could further decrease the battery usage of Syncthing as it should receive a push notification issued by other devices if those want to reach our device. This might be especially useful on Android devices, where Syncthing itself may be effectively stopped until a push message is received.

#### Use GCM

Using Google Cloud Messaging (GCM) was rejected as it makes Syncthing (partially) dependent on Google services, see syncthing/syncthing-android#230. I understand & support this decision.

#### Keep current state

According to [this forum post](https://forum.syncthing.net/t/push-to-android-mobile/13273/2), Syncthing does already implement a push-style protocol. To be able to receive changes in real-time, as of now and assuming behind strict NATs & firewalls, Syncthing needs to maintain at least one connection to somewhere, either to another device or to a public relay server.

A UP distributor app is required to do the same. However, using UP has following advantages:
- UP "unifies" the long-live connections of each app into a single connection, maintained by a single app. Theoretical, one app with one connection is better at saving battery than multiple apps each maintaining their own connection.
- It is easier to keep optimizing a few UP distributor apps in terms of battery usage than each app.
- The protocol/method used between a UP distributor app and its server is independent of the UP spec & independent of the Syncthing protocol, making it easier to optimize & making it a user’s choice.
  - E.g. [NoProvider2Push](https://github.com/NoProvider2Push/android) is a UP distributor which assumes the Android phone can be reached directly using a public IP. This may be established by any means (VPN, port forwardings, sick mobile ISPs, …). This allows users to save more battery life. (I know that Syncthing already does support that, that’s just meant as an example.)
  - theoretically a UP distributor can be implemented by using SMS (no known implementation yet)
  - Users can still opt-in to use GCM in favor of battery power by using [UP-FCM Distributor](https://unifiedpush.org/users/distributors/fcm/), which is not possible with Syncthing as of today.

### (Disclosure)

I’m new to the Syncthing community. I searched the [forum](https://forum.syncthing.net) and the GitHub issue trackers of [the Android app](https://github.com/syncthing/syncthing-android/issues) and [the main repo](https://github.com/syncthing/syncthing/issues). Discussions I found where referenced in this issue.

This FR is open to discussion. I hope this is the right way to discuss such feature request. I proposed it as I did not found any reason in my research which fundamentally speaks against this FR. I can understand if implementing UnifiedPush is deemed too hard.

**Comment by calmh** [comment]
Thanks for taking the time to write this up and file it, I can see you've put a lot of effort in it. Unfortunately, it's extremely unlikely to happen. Adding an Android specific push protocol thing to Syncthing "core" is nowhere on the horizon. Perhaps this could make sense as some sort of companion app to Syncthing paired with an implementation in the Android app/wrapper, I'm not sure.

**Comment by Lcchy** [comment]
I was very coincidentally looking into the high battery usage on Android today and found your issue as I was searching for a more efficient syncing.

I also think that such a system would be a good way to make continuous syncing a more viable use case on Android.
The numerous forum questions and issues about power usage on Android seem to indicate a wider issue.

If it is not realistic to be added to the core, then, as @calmh said, a companion daemon could listen for modifications via the REST Api and send out push notification towards syncthing-android.

I personally don't have any dev experience with Android or syncthing but would be up for helping out if there would be such a project.

**Comment by bt90** [comment]
The least invasive change would be to periodically launch Syncthing using the WorkManager API.

But the first step would need to be finding someone who is willing to invest time in the app.

**Comment by Lcchy** [comment]
A sketchy proof of concept could also maybe use [intents](https://github.com/syncthing/syncthing-android/wiki/Remote-Control-by-Broadcast-Intents) to wake up/stop the app without modifying it?

**Comment by Zocker1999NET** [comment]
> Adding an Android specific push protocol thing to Syncthing "core" is nowhere on the horizon

That feels quite sad. I find it possible that the possibility to initiate connections via a HTTPS callback (like UP does) could become handy for non-Android devices as well. For example [UP is defined for D-Bus as well](https://unifiedpush.org/developers/spec/dbus/), through not many apps support that as of now. And
However, I can understand the resistance against a, possibly breaking, major protocol change.
So I’m asking for planning reasons: If a PR with that implemented would come by, do you think that change would to be accepted? Or do you think it is just highly unlikely that an existing dev would be ready to implement that?


> as some sort of companion app to Syncthing

> A sketchy proof of concept could also maybe use [intents](https://github.com/syncthing/syncthing-android/wiki/Remote-Control-by-Broadcast-Intents) to wake up/stop the app without modifying it?

That seems a good idea for a proof of concept. I haven’t had time to look into the BEP, so I don’t know if the a device tracks what it still has to send to each other device or not. If not, implementing such a companion app seems hard as it then needs to track that info from the outside. Otherwise how should it determine if the Android device is (probably) missing some infos, esp. after the PC device restarted?

**Comment by calmh** [comment]
> If a PR with that implemented would come by, do you think that change would to be accepted? Or do you think it is just highly unlikely that an existing dev would be ready to implement that?

I find it hard to envision what precisely is suggested. For any PR the process is roughly the same, seeing what's the tradeoff in complexity, maintainability, usefulness, etc. A complex change for some niche use case? Probably not going to happen. A small addition that doesn't change the core of things? Sure, why not. I can't speak for other devs about their likelihood of implementing anything, for me it's zero though. :)

**Comment by calmh** [comment]
Seems to me like you really only need the "something changed locally" event from Syncthing (which already exists) and then send a push notification of some kind to wake up the other end. No need to be detailed about what changed why or tie into the protocol; they'll sort that out once they talk due to the wakeup. 🤷

**Comment by Zocker1999NET** [comment]
> I find it hard to envision what precisely is suggested so I can't predict. For any PR the process is roughly the same

Okay, thanks for the answer. Then I may look into that when I have some spare time™.

> Seems to me like you really only need the "something changed locally" event from Syncthing (which already exists)

Essentially, when everything is running smoothly, yes. I was already a little bit over-invested in edge cases (if the device with the companion is restarted, or if the companion app crashes, or how to handle cases where both devices are just sometimes online at the same time).

> I don't see why any deeper involvement would be required on the Syncthing core side.

From a user perspective, an external companion / add-on seems harder to install & configure, esp. when UP may not require further Syncthing-specific configuration if its support would be natively integrated into the core. But I can understand your arguments from a dev perspective.

When people have time, IMO a companion add-on as a proof of concept could be implemented to check out if & how good the idea works.

**Comment by agx** [comment]
Note that this wouldn't only be useful on Android. Mainline Linux based phones running a Phosh/GNOME/KDE/SXMO or similar would also benefit.
